File: ios/QadamLastPuff/Data/PreferencesManager.swift
import Foundation

final class PreferencesManager {
    private let defaults = UserDefaults.standard

    private enum Keys {
        static let isFirstLaunch = "is_first_launch"
        static let onboardingCompleted = "onboarding_completed"
        static let darkTheme = "dark_theme"
        static let notificationsEnabled = "notifications_enabled"
        static let notificationHour = "notification_hour"
        static let notificationMinute = "notification_minute"
        static let unlockedAchievements = "unlocked_achievements"
        static let totalCoins = "total_coins"
        static let longestStreak = "longest_streak"
        static let familyPhotoUri = "family_photo_uri"
        static let personalLetter = "personal_letter"
        static let recoveryIndex = "recovery_index"
        static let lastPersonalLetterDay = "last_personal_letter_day"
    }

    var isFirstLaunch: Bool {
        get {
            if defaults.object(forKey: Keys.isFirstLaunch) == nil { return true }
            return defaults.bool(forKey: Keys.isFirstLaunch)
        }
        set { defaults.set(newValue, forKey: Keys.isFirstLaunch) }
    }

    var onboardingCompleted: Bool {
        get { defaults.bool(forKey: Keys.onboardingCompleted) }
        set { defaults.set(newValue, forKey: Keys.onboardingCompleted) }
    }

    var darkTheme: Bool? {
        get {
            guard defaults.object(forKey: Keys.darkTheme) != nil else { return nil }
            return defaults.bool(forKey: Keys.darkTheme)
        }
        set {
            if let newValue { defaults.set(newValue, forKey: Keys.darkTheme) }
            else { defaults.removeObject(forKey: Keys.darkTheme) }
        }
    }

    var notificationsEnabled: Bool {
        get { defaults.object(forKey: Keys.notificationsEnabled) as? Bool ?? true }
        set { defaults.set(newValue, forKey: Keys.notificationsEnabled) }
    }

    var notificationHour: Int {
        get { defaults.object(forKey: Keys.notificationHour) as? Int ?? 9 }
        set { defaults.set(newValue, forKey: Keys.notificationHour) }
    }

    var notificationMinute: Int {
        get { defaults.object(forKey: Keys.notificationMinute) as? Int ?? 0 }
        set { defaults.set(newValue, forKey: Keys.notificationMinute) }
    }

    var unlockedAchievements: Set<String> {
        get {
            let raw = defaults.string(forKey: Keys.unlockedAchievements) ?? ""
            return Set(raw.split(separator: ",").map(String.init).filter { !$0.isEmpty })
        }
        set {
            defaults.set(newValue.sorted().joined(separator: ","), forKey: Keys.unlockedAchievements)
        }
    }

    var totalCoins: Int {
        get { defaults.integer(forKey: Keys.totalCoins) }
        set { defaults.set(newValue, forKey: Keys.totalCoins) }
    }

    var longestStreak: Int {
        get { defaults.integer(forKey: Keys.longestStreak) }
        set { defaults.set(newValue, forKey: Keys.longestStreak) }
    }

    var familyPhotoUri: String? {
        get { defaults.string(forKey: Keys.familyPhotoUri) }
        set { defaults.set(newValue, forKey: Keys.familyPhotoUri) }
    }

    var personalLetter: String? {
        get { defaults.string(forKey: Keys.personalLetter) }
        set { defaults.set(newValue, forKey: Keys.personalLetter) }
    }

    var recoveryIndexRaw: String? {
        get { defaults.string(forKey: Keys.recoveryIndex) }
        set { defaults.set(newValue, forKey: Keys.recoveryIndex) }
    }

    var lastPersonalLetterDay: Int {
        get { defaults.integer(forKey: Keys.lastPersonalLetterDay) }
        set { defaults.set(newValue, forKey: Keys.lastPersonalLetterDay) }
    }

    func canShowPersonalLetterToday() -> Bool {
        let today = Int(Date().timeIntervalSince1970 / 86400)
        return lastPersonalLetterDay != today
    }

    func markPersonalLetterShownToday() {
        lastPersonalLetterDay = Int(Date().timeIntervalSince1970 / 86400)
    }

    func addCoin() { totalCoins += 1 }

    func unlockAchievement(_ id: String) {
        var set = unlockedAchievements
        set.insert(id)
        unlockedAchievements = set
    }

    func clearAll() {
        let keys = [
            Keys.unlockedAchievements, Keys.totalCoins, Keys.longestStreak,
            Keys.familyPhotoUri, Keys.personalLetter, Keys.recoveryIndex
        ]
        keys.forEach { defaults.removeObject(forKey: $0) }
    }
}

File: app/src/main/java/com/qadam/lastpuff/data/datastore/PreferencesManager.kt
package com.qadam.lastpuff.data.datastore

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "qadam_prefs")

class PreferencesManager(context: Context) {
    private val dataStore = context.dataStore

    private object Keys {
        val IS_FIRST_LAUNCH = booleanPreferencesKey("is_first_launch")
        val ONBOARDING_COMPLETED = booleanPreferencesKey("onboarding_completed")
        val DARK_THEME = booleanPreferencesKey("dark_theme")
        val NOTIFICATIONS_ENABLED = booleanPreferencesKey("notifications_enabled")
        val NOTIFICATION_HOUR = intPreferencesKey("notification_hour")
        val NOTIFICATION_MINUTE = intPreferencesKey("notification_minute")
        val QUOTE_INDEX = intPreferencesKey("quote_index")
        val QUOTE_DATE = longPreferencesKey("quote_date")
        val UNLOCKED_ACHIEVEMENTS = stringPreferencesKey("unlocked_achievements")
        val LONGEST_STREAK = intPreferencesKey("longest_streak")
        val TOTAL_COINS = intPreferencesKey("total_coins")
        val FAMILY_PHOTO_URI = stringPreferencesKey("family_photo_uri")
        val PERSONAL_LETTER = stringPreferencesKey("personal_letter")
        val RECOVERY_INDEX = stringPreferencesKey("recovery_index")
    }

    val isFirstLaunch: Flow<Boolean> =
        dataStore.data.map { it[Keys.IS_FIRST_LAUNCH] ?: true }

    val onboardingCompleted: Flow<Boolean> =
        dataStore.data.map { it[Keys.ONBOARDING_COMPLETED] ?: false }
    val darkTheme: Flow<Boolean?> = dataStore.data.map { it[Keys.DARK_THEME] }
    val notificationsEnabled: Flow<Boolean> =
        dataStore.data.map { it[Keys.NOTIFICATIONS_ENABLED] ?: true }
    val notificationHour: Flow<Int> = dataStore.data.map { it[Keys.NOTIFICATION_HOUR] ?: 9 }
    val notificationMinute: Flow<Int> = dataStore.data.map { it[Keys.NOTIFICATION_MINUTE] ?: 0 }
    val familyPhotoUri: Flow<String?> = dataStore.data.map { it[Keys.FAMILY_PHOTO_URI] }
    val personalLetter: Flow<String?> = dataStore.data.map { it[Keys.PERSONAL_LETTER] }
    val recoveryIndexRaw: Flow<String?> = dataStore.data.map { it[Keys.RECOVERY_INDEX] }

    fun quoteIndexFlow(): Flow<Int> = dataStore.data.map { it[Keys.QUOTE_INDEX] ?: 0 }
    fun quoteDateFlow(): Flow<Long> = dataStore.data.map { it[Keys.QUOTE_DATE] ?: 0L }
    fun unlockedAchievementsFlow(): Flow<Set<String>> = dataStore.data.map {
        (it[Keys.UNLOCKED_ACHIEVEMENTS] ?: "").split(",").filter { s -> s.isNotBlank() }.toSet()
    }

    fun totalCoinsFlow(): Flow<Int> = dataStore.data.map { it[Keys.TOTAL_COINS] ?: 0 }
    fun longestStreakFlow(): Flow<Int> = dataStore.data.map { it[Keys.LONGEST_STREAK] ?: 0 }

    suspend fun setFirstLaunchComplete() {
        dataStore.edit { it[Keys.IS_FIRST_LAUNCH] = false }
    }

    suspend fun setOnboardingCompleted(completed: Boolean) {
        dataStore.edit { it[Keys.ONBOARDING_COMPLETED] = completed }
    }

    suspend fun setDarkTheme(enabled: Boolean) {
        dataStore.edit { it[Keys.DARK_THEME] = enabled }
    }

    suspend fun setNotificationsEnabled(enabled: Boolean) {
        dataStore.edit { it[Keys.NOTIFICATIONS_ENABLED] = enabled }
    }

    suspend fun setNotificationTime(hour: Int, minute: Int) {
        dataStore.edit {
            it[Keys.NOTIFICATION_HOUR] = hour
            it[Keys.NOTIFICATION_MINUTE] = minute
        }
    }

    suspend fun setQuoteIndex(index: Int, date: Long) {
        dataStore.edit {
            it[Keys.QUOTE_INDEX] = index
            it[Keys.QUOTE_DATE] = date
        }
    }

    suspend fun unlockAchievement(id: String) {
        dataStore.edit { prefs ->
            val current = (prefs[Keys.UNLOCKED_ACHIEVEMENTS] ?: "").split(",")
                .filter { it.isNotBlank() }.toMutableSet()
            current.add(id)
            prefs[Keys.UNLOCKED_ACHIEVEMENTS] = current.joinToString(",")
        }
    }

    suspend fun addCoins(amount: Int) {
        if (amount <= 0) return
        dataStore.edit { prefs ->
            prefs[Keys.TOTAL_COINS] = (prefs[Keys.TOTAL_COINS] ?: 0) + amount
        }
    }

    suspend fun setLongestStreak(days: Int) {
        dataStore.edit { prefs ->
            val current = prefs[Keys.LONGEST_STREAK] ?: 0
            if (days > current) prefs[Keys.LONGEST_STREAK] = days
        }
    }

    suspend fun setFamilyPhotoUri(uri: String?) {
        dataStore.edit {
            if (uri == null) it.remove(Keys.FAMILY_PHOTO_URI)
            else it[Keys.FAMILY_PHOTO_URI] = uri
        }
    }

    suspend fun setRecoveryIndex(encoded: String) {
        dataStore.edit { it[Keys.RECOVERY_INDEX] = encoded }
    }

    suspend fun setPersonalLetter(letter: String?) {
        dataStore.edit {
            if (letter.isNullOrBlank()) it.remove(Keys.PERSONAL_LETTER)
            else it[Keys.PERSONAL_LETTER] = letter
        }
    }

    suspend fun clearAll() {
        dataStore.edit { it.clear() }
    }
}
